core/server/common/src/main/java/alluxio/master/journal/raft/DownloadObserver.java

```diff
@@ -0,0 +1,205 @@
+
+  /***
+   * Builds a stream for leader to download a snapshot.
+   *
+   * @param storage the snapshot storage
+   * @param stream the response stream
+   * @return the download stream for leader
+   */
+  public static DownloadObserver<UploadSnapshotPResponse, UploadSnapshotPRequest> forLeader(
+      SimpleStateMachineStorage storage, StreamObserver<UploadSnapshotPResponse> stream) {
+    return new DownloadObserver<>(storage,
+        offset -> UploadSnapshotPResponse.newBuilder().setOffsetReceived(offset).build(),
+        UploadSnapshotPRequest::getData, stream);
+  }
+
+  /***
+   * Builds a stream for follower to download a snapshot.
+   *
+   * @param storage the snapshot storage
+   * @return the download stream for follower
+   */
+  public static DownloadObserver<DownloadSnapshotPRequest, DownloadSnapshotPResponse>
+      forFollower(SimpleStateMachineStorage storage) {
+    return new DownloadObserver<>(storage,
+        offset -> DownloadSnapshotPRequest.newBuilder().setOffsetReceived(offset).build(),
+        DownloadSnapshotPResponse::getData, null);
+  }
+
+  private DownloadObserver(SimpleStateMachineStorage storage, Function<Long, S> messageBuilder,
+      Function<R, SnapshotData> dataGetter, StreamObserver<S> stream) {
+    mStorage = storage;
+    mMessageBuilder = messageBuilder;
+    mDataGetter = dataGetter;
+    mStream = stream;
+  }
+
+  @Override
+  public void onNext(R response) {
+    try {
+      onNextInternal(response);
+    } catch (Exception e) {
+      mStream.onError(e);
+      mFuture.completeExceptionally(e);
+      cleanup();
+    }
+  }
+
+  private void cleanup() {
+    if (mOutputStream != null) {
+      try {
+        mOutputStream.close();
+      } catch (IOException ioException) {
+        LOG.error("Error closing snapshot file", ioException);
+      }
+    }
+    if (mTempFile != null && !mTempFile.delete()) {
+      LOG.error("Error deleting snapshot file {}", mTempFile.getPath());
+    }
+  }
+
+  private void onNextInternal(R response) throws IOException {
+    TermIndex termIndex = TermIndex.newTermIndex(
+        mDataGetter.apply(response).getSnapshotTerm(),
+        mDataGetter.apply(response).getSnapshotIndex());
+    if (mTermIndex == null) {
+      // new start, check if there is already a download
+      LOG.info("Downloading new snapshot {}", termIndex);
+      mTermIndex = termIndex;
+      // start a new file
+      mTempFile = RaftJournalUtils.createTempSnapshotFile(mStorage);
+
+      mTempFile.deleteOnExit();
+      mStream.onNext(mMessageBuilder.apply(0L));
+    } else {
+      if (!termIndex.equals(mTermIndex)) {
+        throw new IOException(String.format(
+            "Mismatched term index when downloading the snapshot. expected: %s actual: %s",
+            mTermIndex, termIndex));
+      }
+      if (!mDataGetter.apply(response).hasChunk()) {
+        throw new IOException(String.format("A chunk is missing from the response %s", response));
```

Comment:
include the snapshot id/file?

Reply:
Added local file path. Snapshot id is included in the response.